src/environment.rs

```diff
@@ -4,103 +4,133 @@ use crate::interpreter::scheme;
-            definitions: scheme::base::base_library::<R, StandardEnv<R>>(),
+            definitions: RefCell::new(scheme::base::base_library::<R, StandardEnv<R>>()),
         }
     }
-    fn new_child(parent: Rc<RefCell<StandardEnv<R>>>) -> Self {
+    fn new_child(parent: Rc<StandardEnv<R>>) -> Self {
         Self {
             parent: Some(parent),
-            definitions: HashMap::new(),
+            definitions: RefCell::new(HashMap::new()),
         }
     }
 
-    fn define(&mut self, name: String, value: Value<R, Self>) {
-        self.definitions.insert(name, value);
+    fn define(&self, name: String, value: Value<R, Self>) {
+        self.definitions.borrow_mut().insert(name, value);
     }
 
-    fn get(&self, name: &str) -> Option<Value<R, Self>> {
-        match self.definitions.get(name) {
-            None => match &self.parent {
-                None => return None,
-                Some(parent) => parent.borrow().get(name),
-            },
-            Some(value) => Some(value.clone()),
+    fn get(&self, name: &str) -> Option<Ref<Value<R, Self>>> {
+        if self.definitions.borrow().contains_key(name) {
+            Some(Ref::map(self.definitions.borrow(), |definitions| {
+                definitions.get(name).unwrap()
+            }))
+        } else {
+            match &self.parent {
+                Some(parent) => parent.get(name),
+                None => None,
+            }
+        }
+    }
+    fn get_mut(&self, name: &str) -> Option<RefMut<Value<R, Self>>> {
+        if self.definitions.borrow().contains_key(name) {
+            Some(RefMut::map(self.definitions.borrow_mut(), |definitions| {
+                definitions.get_mut(name).unwrap()
+            }))
+        } else {
+            match &self.parent {
+                Some(parent) => parent.get_mut(name),
+                None => None,
+            }
         }
     }
 
-    fn set(&mut self, name: &str, value: Value<R, Self>) -> Result<(), SchemeError> {
-        match self.definitions.get_mut(name) {
+    fn set(&self, name: &str, value: Value<R, Self>) -> Result<(), SchemeError> {
+        match self.definitions.borrow_mut().get_mut(name) {
             None => match &self.parent {
                 None => logic_error!("unbound variable {}", name),
-                Some(parent) => parent.borrow_mut().set(name, value)?,
+                Some(parent) => parent.set(name, value)?,
             },
             Some(variable) => *variable = value,
         };
         Ok(())
     }
 
-    fn iter_local_definitions<'a>(
-        &'a self,
-    ) -> Box<dyn 'a + Iterator<Item = (&'a String, &'a Value<R, Self>)>> {
-        Box::new(self.definitions.iter())
+    fn iter_local_definitions<'a, 'b: 'a>(&'b self) -> RefVal<'a, DefinitionIter<'b, R, Self>> {
+        Ref::map_val(self.definitions.borrow(), |definitions| {
+            let iter: DefinitionIter<'b, R, Self> =
+                Box::new(definitions.iter().map(|(s, v)| (s.as_str(), v)));
```

Comment:
as_str is not necessary

Reply:
Because the return iterator is for item type ``(&str, &Value)``, not for ``(&String, &Value)``, so ``as_str`` is required.
I think ``&str`` is more general than ``&String``, so I change the item type.

Reply:
Currently just return what (&key, &value) stores is enough. We will consider it later in name interning